### ISSUE TITLE
[Discussion] choose appropriate tool for WCAG auto testing / visual regression tests

### ISSUE THREAD

**Issue by davojta** [wagtail/wagtail#4871]
Problem:
*  Wagtail need a tool for auto frontend tests
* wagtail need auto testing for wcag

Here is the cypress test for wagtail - https://github.com/davojta/test-cypress-for-wagtail

There are several solutions for wcag testing:

* pa11yjs
* axe-core

There also some options for visual regression testing

* cypress
* web driver
* puppeteer 

To be updated later )

**Comment by thibaudcolas** [comment]
To add a bit more context, here are the things we want to test accessibility-wise at the moment:

- Icon font replacement to SVG, https://github.com/wagtail/wagtail/issues/4199#issuecomment-358640129, #4821
- Document outline & landmarks,  https://github.com/wagtail/wagtail/issues/4199#issuecomment-358640129
- Textual labels to icon-only elements, https://github.com/wagtail/wagtail/issues/4199#issuecomment-358640129
- Focus management, https://github.com/wagtail/wagtail/issues/4199#issuecomment-358640129
- Contrast ratio (WCAG 2.0 AA level), #2224, #4788 

For the actual accessibility tester, I feel pretty strongly about picking aXe since it's battle-tested, and is know to have a "no false positives" policy for use in automated tests.

**Comment by thibaudcolas** [comment]
Looking into this further, I see the need for at least two tools / sets of tools, for two separate scenarios:

- Dev-time "static analysis" / linting, that can give developers some feedback as they work on their change (e.g. a PR to add a new view in the admin). And can also run through the source code in CI.
- Automated E2E integration testing, with browser automation tools, that runs a more comprehensive set of checks.

For the E2E accessibility tests, the most mature is [pa11y](https://github.com/pa11y/pa11y), already mentioned above. They are working on an Axe integration that would make it perfect (https://github.com/pa11y/pa11y/pull/408). If this doesn't work out, I think the next best would be to have our own Puppeteer + Axe integration (pa11y is an abstraction layer for Puppeteer interactions).

For the dev-time static analysis, I haven't found that many options. Common HTML linting tools do not have good support for fragments, or templating syntax, and do not have mature rules ecosystems like ESLint/stylelint/flake8.

That said [jinjalint](https://github.com/motet-a/jinjalint) looks promising. It works out of the box, and finds issues in Wagtail's codebase already (16). In the React world, there's https://github.com/evcohen/eslint-plugin-jsx-a11y, which I think is already used in Wagtail's React linting.

Finally, there is a third category of tools which seem useful during development: runtime analysers that seem like they could be hooked into the dev environment, as well as unit or integration tests:

- https://github.com/dequelabs/react-axe for React code
- https://github.com/peterbe/django-html-validator for Django templates. See also https://stackoverflow.com/questions/3086637/how-should-i-validate-html-in-django-templates-during-the-development-process.

Feedback most welcome!

**Comment by thibaudcolas** [comment]
Relevant PR: #5245. This does not contain automated integration tests running in CI, because we first need to decide what would be a proper testing environment for those (bakerydemo, a demo site within the Wagtail repository, or something else entirely).

Current progress for the automated integration testing with Pa11y / Puppeteer / jest / Axe is over at https://github.com/thibaudcolas/wagtail-tooling.

**Comment by thibaudcolas** [comment]
@Scotchester interested in looking into this if others want to join forces.

**Comment by thibaudcolas** [comment]
From @allcaps – we can use django-html-validator with an official validator image, which makes the setup _much_ simpler:

```
docker run -d -p 8888:8888 validator/validator:latest
```

Image: https://hub.docker.com/r/validator/validator/
